File: problems/859-buddy-strings/solution.java
class Solution {
    public boolean buddyStrings(String s, String goal) {
        if (s.length() != goal.length()) return false;
        if (s.equals(goal)) {
            int[] count = new int[26];
            for (char c : s.toCharArray()) {
                count[c - 'a']++;
                if (count[c - 'a'] > 1) return true;
            }
            return false;
        }
        int diff = 0;
        int i = 0;
        int j = 0;
        for (; i < s.length(); i++) {
            if (s.charAt(i) != goal.charAt(i)) {
                diff++;
                j = i;
            }
            if (diff > 2) return false;
        }
        
        if (diff == 2) {
            return s.charAt(i - 1) == goal.charAt(j) && s.charAt(j) == goal.charAt(i -1);
        }
        return false;

    }
}

File: problems/859-buddy-strings/solution.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False
        if s == goal:
            seen = set()
            for char in s:
                if char in seen:
                    return True
                seen.add(char)
            return False
        diff = []
        for i in range(len(s)):
            if s[i] != goal[i]:
                diff.append(i)
        if len(diff) == 2:
            i, j = diff
            return s[i] == goal[j] and s[j] == goal[i]
        return False